packages/composites/the-graph/src/methods/prices/dex/curve.ts

```diff
@@ -0,0 +1,164 @@
+			query: getTokenPoolQuery,
+			variables: {
+				symbol
+			},
+			graphqlEndpoint: this.url
+		}
+		const response = await fetchFromGraphqlAdapter(jobRunID, data)
+		if (!response.result.data.tokens.length) {
+			const error = response.result.error || "Failed to get token information, enter the exact Token symbol"
+			throw new Error(error)
+		}
+		const tokensPolls = response.result.data.tokens[0].coins
+		const pools = tokensPolls.map((coin: any) => {
+			return coin.pool.name
+		})
+		return pools
+	}
+
+	async getPoolPrice(jobRunID: string, inputParams: DexQueryInputParams): Promise < number > {
+		const {
+			baseCoinTicker,
+			quoteCoinTicker,
+			referenceContract,
+			referenceContractDivisor
+		} = inputParams
+		const tokenPollsName = await this.getTokenPoolName(jobRunID, quoteCoinTicker)
+		Logger.info(`Found ${tokenPollsName.length} pools`)
+		const balances = await Promise.all(tokenPollsName.map(async (name: string) => {
+			const pools = await this.getPool(jobRunID, name);
+			Logger.info(`Fetching balance from '${name}' pool`)
+			let token0Balance
+			let token1Balance
+			pools.underlyingCoins.map((pool: UnderlyingCoins) => {
+				if (quoteCoinTicker === pool.token.symbol) {
+					token0Balance = pool.balance
+				}
+				if (baseCoinTicker === pool.token.symbol) {
+					token1Balance = pool.balance
+				}
+			})
+			if (token0Balance === undefined || token1Balance === undefined) {
+				Logger.info(`'${name}' pool doesn't have both tokens`)
+				return null
+			}
+			return {
+				token0Balance,
+				token1Balance
+			}
+		}))
+		const pool = balances.filter(pool => pool) as PoolBalance[]
+		if (!pool.length) {
+			throw new Error(`There is not a ${quoteCoinTicker}/${baseCoinTicker} pool`)
+		}
+		const token0Balance = pool[0].token0Balance
+		const token1Balanace = pool[0].token1Balance
+		const poolUnderlyingPrice = token1Balanace / token0Balance
+		Logger.info(`Price of ${quoteCoinTicker}/${baseCoinTicker} is ${poolUnderlyingPrice}`)
+		if(!referenceContract) {
+			return poolUnderlyingPrice
+		}
+		const mofifiedPrice  = await this.modifyResultByFeedResult(inputParams, referenceContractDivisor)
+		return mofifiedPrice
+	}
+
+	async modifyResultByFeedResult (
+		inputParams: DexQueryInputParams,
+		currentPrice: number,
+	): Promise < number > {
+		const {
+			baseCoinTicker,
+			quoteCoinTicker,
+			referenceContract,
+			referenceContractDivisor,
+			referenceModifierAction
+		} = inputParams
+
+		Logger.info(
+			`Price of ${quoteCoinTicker}/${baseCoinTicker} is going to be modified by the result returned from ${referenceContract} by ${referenceContractDivisor}`,
+		)
+		const modifierTokenPrice = await getLatestAnswer(referenceContract, referenceContractDivisor)
+		Logger.info(`Feed ${referenceContract} returned a value of ${modifierTokenPrice}`)
+		if (referenceModifierAction === ReferenceModifierAction.DIVIDE) {
+			return currentPrice / modifierTokenPrice
+		}
+		return currentPrice * modifierTokenPrice
+	}
```

Comment:
It seems we are duplicating this functionality on both Uniswap and Curve. Can we instead handle this in a single place to avoid duplication for each DEX?